## linear regression, Ridge and Lasso regression and Logistic Regression:

In [2]:
#dataset-house pricing dataset(prsent in slearn)
#To import datset:
from sklearn.datasets import fetch_california_housing

In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

%matplotlib inline command is a magic function used in Jupyter notebooks and IPython environments. It sets up the environment to display plots and figures created with Matplotlib directly within the notebook or IPython cell output.

With %matplotlib inline, any plots generated by Matplotlib will be displayed inline within the notebook or IPython environment, directly below the code cell that generates them. This is useful for interactive data analysis and visualization within the notebook.

plots are rendered as static images. If you need interactive plots, you might use %matplotlib notebook or %matplotlib widget instead

In [4]:
data_frame = fetch_california_housing()
#To load datset- using fetch_california_housing library 
#just intializing it

In [5]:
type(data_frame)

sklearn.utils._bunch.Bunch

In [6]:
data_frame

#This is in the form of key value pairs :
#data
#target
#feature names
#so we need to combine these in proper way-in data frame.

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

In [7]:
pd.DataFrame(data_frame.data)
#Converting entrily to a data frame.
#.data-all the values of features
#we will get entire data set-f1,f2,  feature 12

0     1         2         3       4         5      6       7
0      8.3252  41.0  6.984127  1.023810   322.0  2.555556  37.88 -122.23
1      8.3014  21.0  6.238137  0.971880  2401.0  2.109842  37.86 -122.22
2      7.2574  52.0  8.288136  1.073446   496.0  2.802260  37.85 -122.24
3      5.6431  52.0  5.817352  1.073059   558.0  2.547945  37.85 -122.25
4      3.8462  52.0  6.281853  1.081081   565.0  2.181467  37.85 -122.25
...       ...   ...       ...       ...     ...       ...    ...     ...
20635  1.5603  25.0  5.045455  1.133333   845.0  2.560606  39.48 -121.09
20636  2.5568  18.0  6.114035  1.315789   356.0  3.122807  39.49 -121.21
20637  1.7000  17.0  5.205543  1.120092  1007.0  2.325635  39.43 -121.22
20638  1.8672  18.0  5.329513  1.171920   741.0  2.123209  39.43 -121.32
20639  2.3886  16.0  5.254717  1.162264  1387.0  2.616981  39.37 -121.24

[20640 rows x 8 columns]

In [8]:
#To add target feature:
dataset = pd.DataFrame(data_frame.data)
dataset.columns = data_frame.feature_names
print(dataset.head) #To view 

<bound method NDFrame.head of        MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Lo

Using print() explicitly prints the output of .head() to the standard output. This is useful when you're running code in a script or a standard Python interpreter where you want to ensure that the output is shown in the console or terminal.

When you use .head() without print(), it will display the first few rows of the dataset in an interactive environment, like Jupyter notebooks or IPython consoles, automatically showing the result as the output of the cell or line.

In [9]:
dataset.head()
#5 records of dataset

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [10]:
dataset
#whole data

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
0        -122.23  
1        -122.22  
2        -122.24  
3        -122.25  
4        -122.25  
...          ...  
20635    -121.09  
20636    -121.21  
20637    -121.22  
20638    -121.32  
20639    -121.24  

[20640 rows x 8 columns]

I have taken all independent features.

still have dependent feature.

Will create new feature like price(median) and will assign to target value(median house value for California districts,expressed in hundreds of thousands of dollars ($100,000))

will take it and put it like dependent feature.

    

In [11]:
dataset['MedPrice'] = data_frame.target 

In [12]:
dataset

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedPrice  
0        -122.23     4.526  
1        -122.22     3.585  
2        -122.24     3.521  
3        -122.25     3.413  
4        -122.25     3.422  
...          ...       ...  
20635    -121.09     0.781  
20636    -121.21     0.771  
20637    -121.22     0.923  
20638    -121.32     0.847  
20639    -121.24     0.894  

[20640 rows x 9 columns]

So, here MedInc, HouseAge, AveRooms, AveBedrms, Population, AveOccup, Latitude, Longitude are all independent features.

MedPrice is the dependent feature.

In [13]:
#To solve linear regression need to divide independent and dependent features correctly.
#X independent features.
#y dependent feature.
#iloc - present in dataframe
#To understand the location of independent features-need to skip the last column or feature
X = dataset.iloc[:,:-1] #all the rowsa nd columns except for last
y = dataset.iloc[:,-1] # all rows of last column
#Dividing the dataset into train and test.

In [14]:
X

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
0        -122.23  
1        -122.22  
2        -122.24  
3        -122.25  
4        -122.25  
...          ...  
20635    -121.09  
20636    -121.21  
20637    -121.22  
20638    -121.32  
20639    -121.24  

[20640 rows x 8 columns]

In [15]:
y

0        4.526
1        3.585
2        3.521
3        3.413
4        3.422
         ...  
20635    0.781
20636    0.771
20637    0.923
20638    0.847
20639    0.894
Name: MedPrice, Length: 20640, dtype: float64

## Algorithm: Linear Regression

need to use few linear regression libraries.

2
- linear regression library from sklearn
- import cross val score(going to do lot of train and test split)



In [16]:
#CROSS Validation:

#Entire dataset - 100 records

#  ______________________________
#|                               |
 # ______________________________

#5 cross validation

#1st CV: 

#test data and remaining all training data

 #______________________________
#| Test |        Training       |
 # _____________________________

#2nd CV:

 #______________________________
#| Test | Test |  Training       |
 # _____________________________

#3rd CV:

#...

#so, will do cross validation 5 times taking diff combination of train and test data.

In [17]:
#linear regression
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score

#intialize
lin_reg = LinearRegression()
#Applying cross validation
#CV- we divide train and test in such a way that every combination of train and 
#test is taken by the model and who's ever accuracy is better than entire thing is combined.

mse = cross_val_score(lin_reg, X, y, scoring='neg_mean_squared_error', cv=5)
#parameters-model-linear reg, X, y
#we are not doing train test split specifically here.
#we are giving entire X and y
#Based on that i'm going to do cross validation
#Can also do train test split intialling and then give x train and y train to do the cross
#validation.


Best practice:
Doing train test split and then only giving train data to do the cross validation.

scoring can be use MSE(mean squared error) and negMSE
can find all of this in sklearn page of cross val score.

Cv - cross validation value
here it is 5.

So, here as we are doiung 5 cross validation mse going to be 5 values.
|

In [18]:
print(mse)
#all diff MSE-as we are doing Cv as 5

[-0.48485857 -0.62249739 -0.64621047 -0.5431996  -0.49468484]


In [19]:
mse_avg = np.mean(mse)
#average of all mean squared errors.

In [20]:
print(mse_avg)

-0.5582901717686543


In [21]:
#Neg as we used -ve mse.
#but considering only mse
#it is 0.558

#This is how we do cross validation.

wrt to linear regression there is not much to modify with the parameter.

So, in order of overcome overfitting and feature selection. 

We use ridge and lasso regression.
    

In [22]:
#To do prediction
lin_reg.fit(X,y)
lin_reg.predict(X)

array([4.13164983, 3.97660644, 3.67657094, ..., 0.17125141, 0.31910524,
       0.51580363])

This line trains (or "fits") the linear regression model (lin_reg) using the input data X and the target values y.
X: This represents the features or independent variables in your dataset.
y: This represents the target or dependent variable you want to predict.

This line uses the trained model (lin_reg) to make predictions on the input data X.
Result: The model produces predicted values based on the patterns it learned during training.

Training the Model: lin_reg.fit(X, y) trains the linear regression model by finding the best-fit line that minimizes the error between the predicted values and the actual values.

Making Predictions: lin_reg.predict(X) uses the trained model to predict the target values for the given input features.

In [23]:
error_val = dataset.iloc[:,-1]-lin_reg.predict(X)

In [24]:
error_val

0        0.394350
1       -0.391606
2       -0.155571
3        0.171402
4        1.008413
           ...   
20635    0.647344
20636    0.217472
20637    0.751749
20638    0.527895
20639    0.378196
Name: MedPrice, Length: 20640, dtype: float64

In [25]:
error_val_avg = np.mean(error_val)

In [26]:
error_val_avg

-7.686530137621859e-15

## Ridge Regression:

In [27]:
#hyperparameter tuning in ridge regression.

#2 diff libraries:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
#This is for doing hyper parameter tuning.(by GridSearchCV)



#in order to get good value - we do hyperparameter tuning.

#model:
ridge = Ridge()

#opening sklearn ridge
#parameters
#1. alpha - learning rate (best parameter to perform hyperparameter tuning)

#2. max iterations - how many times to change theta1 value to get the right value.


In [28]:
#Defining paramters:
#in dictionaries:
params = {'alpha':[1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,20]}

#in grid search CV - they take combination of alpha value and wherever the model performs well
# and gives it as selected bestfit parameter.

ridge_regressor = GridSearchCV(ridge, params, scoring='neg_mean_squared_error', cv=7)
#parameters: model, parameters defined, scoring(neg mean squared error)


In [29]:
ridge_regressor.fit(X, y)
#Can also do train test split and can do this process of model fit on train X and train y.


GridSearchCV(cv=7, estimator=Ridge(),
             param_grid={'alpha': [1e-15, 1e-10, 1e-08, 0.001, 0.01, 1, 5, 10,
                                   20]},
             scoring='neg_mean_squared_error')

In [30]:
print(ridge_regressor.best_params_)  #This gives best parameter(for alpha)
print(ridge_regressor.best_score_)  #This gives best score(error best)

{'alpha': 20}
-0.5548535736454416


bcz of ridge regression. Negative mean square error has definetely became better.

-0.5582901717686543 - for linear regression
    
-0.5548535736454416 - for ridge regression

So, should select linear regression as it performing better.

but, ridge reduce overfitting.(cannot use as performance is becoming very bad)

Now , we can try lasso regression.


## Lasso Regression:

In [31]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV

lasso = Lasso() #intialization of model.

params_l = {'alpha':[1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,20]} #defining parameters.

#cross validation:
lasso_reg = GridSearchCV(lasso, params_l, scoring='neg_mean_squared_error', cv=7)

#fitting the model:
lasso_reg.fit(X, y)


/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.917e+01, tolerance: 2.390e+00
  model = cd_fast.enet_coordinate_descent(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.027e+03, tolerance: 2.268e+00
  model = cd_fast.enet_coordinate_descent(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.770e+03, toler

GridSearchCV(cv=7, estimator=Lasso(),
             param_grid={'alpha': [1e-15, 1e-10, 1e-08, 0.001, 0.01, 1, 5, 10,
                                   20]},
             scoring='neg_mean_squared_error')

In [32]:
print(lasso_reg.best_params_)
print(lasso_reg.best_score_)

{'alpha': 1e-15}
-0.5550162785002432


-0.5582901717686543 - for linear regression 

-0.5548535736454416 - for ridge regression

-0.5550162785002432 - for lasso regression

So, still can focus on linear regression.

In [33]:
#Now by adding more parameters:
lasso_two = Lasso() #intialization of model.

params_l_two = {'alpha':[1e-25,1e-20,1e-17,1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,20,30,35,40,45,50,55,100]} #defining parameters.

#cross validation:
lasso_reg_two = GridSearchCV(lasso_two, params_l_two, scoring='neg_mean_squared_error', cv=9)

#fitting the model:
lasso_reg_two.fit(X, y)


/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.919e+03, tolerance: 2.507e+00 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV instead.
  model = cd_fast.enet_coordinate_descent(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.826e+03, tolerance: 2.426e+00 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.lin

GridSearchCV(cv=9, estimator=Lasso(),
             param_grid={'alpha': [1e-25, 1e-20, 1e-17, 1e-15, 1e-10, 1e-08,
                                   0.001, 0.01, 1, 5, 10, 20, 30, 35, 40, 45,
                                   50, 55, 100]},
             scoring='neg_mean_squared_error')

In [34]:
print(lasso_reg_two.best_params_)

{'alpha': 1e-08}


In [35]:
print(lasso_reg_two.best_score_)

-0.5478151595223992


-0.5582901717686543 - for linear regression

-0.5548535736454416 - for ridge regression

-0.5550162785002432 - for lasso regression 

-0.5478151595223992 - for lasso regression with other parameters.

In [36]:
#Now by adding more parameters:
lasso_thre = Lasso() #intialization of model.

params_l_thre = {'alpha':[1e-8,1e-3,1e-2,1,5,6,7,10,13,14,15,20,21,23,25,30,35,40,45,50,55,60,65,70,75,80,85,90,95,100]} #defining parameters.

#cross validation:
lasso_reg_thre = GridSearchCV(lasso_thre, params_l_thre, scoring='neg_mean_squared_error', cv=10)

#fitting the model:
lasso_reg_thre.fit(X, y)

print(lasso_reg_thre.best_params_)
print(lasso_reg_thre.best_score_)

{'alpha': 1e-08}
-0.5509524298660506


In [37]:
#Need to play with these parameters and take multiple parameters
#which ever gives best result will be taken.


### Now by taking Train and Test split:

In [38]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=45)

#random size can be anything.
#test_size the test will be 25% and train is remaining.


In [39]:
X_train

MedInc  HouseAge   AveRooms  AveBedrms  Population  AveOccup  Latitude  \
14961  5.2741       7.0   6.855372   0.979339       862.0  3.561983     32.76   
11810  4.5150      15.0   6.379061   1.063779      2335.0  2.809868     38.92   
14150  2.6452      18.0   5.108434   1.129088      1149.0  1.977625     32.73   
1236   2.6942      13.0  11.490862   2.164491       924.0  2.412533     38.23   
278    2.6667      50.0   4.746269   1.079602       546.0  2.716418     37.79   
...       ...       ...        ...        ...         ...       ...       ...   
16547  4.0125      20.0   5.538226   1.076453      1191.0  3.642202     37.81   
16317  2.2130      22.0   4.358178   1.078675       781.0  1.616977     38.00   
12895  2.8106      24.0   4.886584   1.024896      1496.0  2.069156     38.65   
6012   3.9784      18.0   4.718954   1.009150      2899.0  3.789542     34.07   
6558   6.8270      36.0   7.021429   1.035714      1897.0  2.710000     34.20   

       Longitude  
14961    -116.92  
11810    -121.07  
14150    -117.07  
1236     -120.37  
278      -122.19  
...          ...  
16547    -121.22  
16317    -121.32  
12895    -121.33  
6012     -117.78  
6558     -118.11  

[15480 rows x 8 columns]

In [40]:
y_train

14961    2.494
11810    1.640
14150    1.542
1236     1.235
278      1.728
         ...  
16547    1.215
16317    0.875
12895    1.832
6012     1.135
6558     3.594
Name: MedPrice, Length: 15480, dtype: float64

In [41]:
X_test

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
3046   2.4044      34.0  5.478632   1.079772      1101.0  3.136752     35.50   
6803   2.5750      40.0  4.329596   1.006726      1367.0  3.065022     34.08   
10112  4.5096      30.0  5.165049   0.993528       845.0  2.734628     33.94   
3126   2.3750       4.0  1.000000   1.000000         6.0  3.000000     35.21   
5008   3.7727      47.0  4.536082   0.890034       653.0  2.243986     34.01   
...       ...       ...       ...        ...         ...       ...       ...   
5461   2.0714      31.0  3.975758   1.139394      1178.0  3.569697     33.99   
11345  4.0313      19.0  4.444444   1.090278      1413.0  3.270833     33.75   
9469   2.7813      36.0  4.650350   0.954545       761.0  2.660839     39.44   
4025   3.8550      18.0  4.770030   1.047478      2897.0  2.149110     34.17   
18082  7.8627      24.0  7.145115   1.018678      1692.0  2.431034     37.26   

       Longitude  
3046     -119.28  
6803     -118.13  
10112    -117.94  
3126     -117.79  
5008     -118.33  
...          ...  
5461     -118.47  
11345    -117.92  
9469     -123.79  
4025     -118.53  
18082    -122.04  

[5160 rows x 8 columns]

In [42]:
y_test

3046     0.65800
6803     2.28400
10112    2.41100
3126     1.37500
5008     1.93000
          ...   
5461     3.00000
11345    1.47500
9469     1.05800
4025     2.43800
18082    5.00001
Name: MedPrice, Length: 5160, dtype: float64

In [43]:
#Algorithm: Linear Regression

linear_reg = LinearRegression()
linear_reg.fit(X_train,y_train)
mse_tt = cross_val_score(linear_reg, X_train, y_train, scoring = 'neg_mean_squared_error', cv=5)
mean_mse_tt = np.mean(mse_tt)
print(mean_mse_tt)

-0.5299054932224389


mean_mse_tt value should be going towards 0.
    
if it goes towards 0 then the performance is better now.



In [44]:
#Algorithm: Ridge Regression

ridge_tt = Ridge()

params_tt = {'alpha':[1e-25,1e-20,1e-17,1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,20,30,35,40,45,50,55,100]}

ridge_reg_tt = GridSearchCV(ridge_tt, params_tt, scoring='neg_mean_squared_error',cv=6)
ridge_reg_tt.fit(X_train,y_train)

GridSearchCV(cv=6, estimator=Ridge(),
             param_grid={'alpha': [1e-25, 1e-20, 1e-17, 1e-15, 1e-10, 1e-08,
                                   0.001, 0.01, 1, 5, 10, 20, 30, 35, 40, 45,
                                   50, 55, 100]},
             scoring='neg_mean_squared_error')

In [45]:
print(ridge_reg_tt.best_params_)  #This gives best parameter(for alpha)
print(ridge_reg_tt.best_score_)  #This gives best score(error best)

{'alpha': 100}
-0.5300117540649317


still the improvement is bad. As not going towards 0.

In [46]:
#Algorithm: Lasso regression
lasso_tt = Lasso() #intialization of model.

params_l_tt = {'alpha':[1e-15,1e-10,1e-8,1e-3,1e-2,1,5,10,20]} #defining parameters.

#cross validation:
lasso_reg_tt = GridSearchCV(lasso_tt, params_l_tt, scoring='neg_mean_squared_error', cv=7)

#fitting the model:
lasso_reg_tt.fit(X_train, y_train)


/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.008e+03, tolerance: 1.769e+00
  model = cd_fast.enet_coordinate_descent(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.046e+03, tolerance: 1.755e+00
  model = cd_fast.enet_coordinate_descent(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:678: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.908e+03, toler

GridSearchCV(cv=7, estimator=Lasso(),
             param_grid={'alpha': [1e-15, 1e-10, 1e-08, 0.001, 0.01, 1, 5, 10,
                                   20]},
             scoring='neg_mean_squared_error')

In [47]:
print(lasso_reg_tt.best_params_)
print(lasso_reg_tt.best_score_)

{'alpha': 0.001}
-0.5292747504107352


In [48]:
#here we can use:
#predict wrt test data
y_pred_lin = linear_reg.predict(X_test)
y_pred_l = lasso_reg_tt.predict(X_test)
y_pred_r = ridge_reg_tt.predict(X_test)

#here we get y_test value as output.

In [49]:
from sklearn.metrics import r2_score

#will use R2 and adjusted R2.

r2_score3=r2_score(y_pred_lin,y_test)
r2_score1=r2_score(y_pred_l,y_test)
r2_score2=r2_score(y_pred_r,y_test)

In [50]:
print(r2_score1)
print(r2_score2)
print(r2_score3)

0.3332111576547441
0.32656007279045185
0.33679522611747614


In [51]:
#linear reg - we can't get 100% as we are drawing straight line.

#So, need to do other algorithms like xgboost, naive bayes.

#Till one limit we can increase the performance after that we can't.



## Logistic Regression:

In [52]:
# can also do grid search cv.

In [53]:
from sklearn.linear_model import LogisticRegression


#for logistic we solve using classification problem
#so, lets take new dataset:

#Dataset:
from sklearn.datasets import load_breast_cancer

In [54]:
data_f = load_breast_cancer()

In [55]:
data_f

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [56]:
dataset_bc = pd.DataFrame(data_f['data'], columns=data_f['feature_names'])

In [57]:
dataset_bc #independent features

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [58]:
dataset_bc['Result']=data_f['target'] #Dependent feature

In [59]:
dataset_bc

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  Result  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

In [60]:
#creating seperate data frames for independent and dependent features.

X = pd.DataFrame(data_f['data'], columns=data_f['feature_names'])

In [61]:
X

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [62]:
y = pd.DataFrame(data_f['target'])

In [63]:
y.columns= ['Result']

In [64]:
y

Result
0         0
1         0
2         0
3         0
4         0
..      ...
564       0
565       0
566       0
567       0
568       1

[569 rows x 1 columns]

In [65]:
#check if the output/dependent feature is balanced or imbalanced:

y['Result'].value_counts()

#it gives count of 1's and 0's

Result
1    357
0    212
Name: count, dtype: int64

In [66]:
#This is probably a balanced dataset.

In [67]:
#Train Test split:

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35, random_state=47)


In [68]:
X_train

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
176        9.904         18.06           64.60      302.4          0.09699   
260       20.310         27.06          132.90     1288.0          0.10000   
206        9.876         17.27           62.92      295.4          0.10890   
319       12.430         17.00           78.60      477.3          0.07557   
519       12.750         16.70           82.51      493.8          0.11250   
..           ...           ...             ...        ...              ...   
59         8.618         11.79           54.34      224.5          0.09752   
23        21.160         23.04          137.20     1404.0          0.09428   
264       17.190         22.07          111.60      928.3          0.09726   
327       12.030         17.93           76.09      446.0          0.07683   
135       12.770         22.47           81.72      506.3          0.09055   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
176           0.12940        0.130700             0.037160         0.1669   
260           0.10880        0.151900             0.093330         0.1814   
206           0.07232        0.017560             0.019520         0.1934   
319           0.03454        0.013420             0.016990         0.1472   
519           0.11170        0.038800             0.029950         0.2120   
..                ...             ...                  ...            ...   
59            0.05272        0.020610             0.007799         0.1683   
23            0.10220        0.109700             0.086320         0.1769   
264           0.08995        0.090610             0.065270         0.1867   
327           0.03892        0.001546             0.005592         0.1382   
135           0.05761        0.047110             0.027040         0.1585   

     mean fractal dimension  ...  worst radius  worst texture  \
176                 0.08116  ...        11.260          24.39   
260                 0.05572  ...        24.330          39.16   
206                 0.06285  ...        10.420          23.22   
319                 0.05561  ...        12.900          20.21   
519                 0.06623  ...        14.450          21.74   
..                      ...  ...           ...            ...   
59                  0.07187  ...         9.507          15.40   
23                  0.05278  ...        29.170          35.59   
264                 0.05580  ...        21.580          29.33   
327                 0.06070  ...        13.070          22.25   
135                 0.06065  ...        14.490          33.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
176            73.07       390.2           0.13010            0.29500   
260           162.30      1844.0           0.15220            0.29450   
206            67.08       331.6           0.14150            0.12470   
319            81.76       515.9           0.08409            0.04712   
519            93.63       624.1           0.14750            0.19790   
..               ...         ...               ...                ...   
59             59.90       274.9           0.17330            0.12390   
23            188.00      2615.0           0.14010            0.26000   
264           140.50      1436.0           0.15580            0.25670   
327            82.74       523.4           0.10130            0.07390   
135            92.04       653.6           0.14190            0.15230   

     worst concavity  worst concave points  worst symmetry  \
176         0.348600               0.09910          0.2614   
260         0.378800               0.16970          0.3151   
206         0.062130               0.05588          0.2989   
319         0.022370               0.02832          0.1901   
519         0.142300               0.08045          0.3071   
..               ...                   ...             ...   
59          0.116800               0.04419         

In [69]:
X_test

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
114        8.726         15.83           55.84      230.9          0.11500   
282       19.400         18.18          127.20     1145.0          0.10370   
252       19.730         19.82          130.70     1206.0          0.10620   
130       12.190         13.29           79.08      455.8          0.10660   
318        9.042         18.90           60.07      244.5          0.09968   
..           ...           ...             ...        ...              ...   
371       15.190         13.21           97.65      711.8          0.07963   
47        13.170         18.66           85.98      534.6          0.11580   
489       16.690         20.20          107.10      857.6          0.07497   
220       13.650         13.16           87.88      568.9          0.09646   
423       13.660         19.13           89.46      575.3          0.09057   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
114           0.08201         0.04132              0.01924         0.1649   
282           0.14420         0.16260              0.09464         0.1893   
252           0.18490         0.24170              0.09740         0.1733   
130           0.09509         0.02855              0.02882         0.1880   
318           0.19720         0.19750              0.04908         0.2330   
..                ...             ...                  ...            ...   
371           0.06934         0.03393              0.02657         0.1721   
47            0.12310         0.12260              0.07340         0.2128   
489           0.07112         0.03649              0.02307         0.1846   
220           0.08711         0.03888              0.02563         0.1360   
423           0.11470         0.09657              0.04812         0.1848   

     mean fractal dimension  ...  worst radius  worst texture  \
114                 0.07633  ...         9.628          19.62   
282                 0.05892  ...        23.790          28.65   
252                 0.06697  ...        25.280          25.59   
130                 0.06471  ...        13.340          17.81   
318                 0.08743  ...        10.060          23.40   
..                      ...  ...           ...            ...   
371                 0.05544  ...        16.200          15.73   
47                  0.06777  ...        15.670          27.95   
489                 0.05325  ...        19.180          26.56   
220                 0.06344  ...        15.340          16.35   
423                 0.06181  ...        15.140          25.50   

     worst perimeter  worst area  worst smoothness  worst compactness  \
114            64.48       284.4            0.1724             0.2364   
282           152.40      1628.0            0.1518             0.3749   
252           159.80      1933.0            0.1710             0.5955   
130            91.38       545.2            0.1427             0.2585   
318            68.62       297.1            0.1221             0.3748   
..               ...         ...               ...                ...   
371           104.50       819.1            0.1126             0.1737   
47            102.80       759.4            0.1786             0.4166   
489           127.30      1084.0            0.1009             0.2920   
220            99.71       706.2            0.1311             0.2474   
423           101.40       708.8            0.1147             0.3167   

     worst concavity  worst concave points  worst symmetry  \
114          0.24560               0.10500          0.2926   
282          0.43160               0.22520          0.3590   
252          0.84890               0.25070          0.2749   
130          0.09915               0.08187          0.3469   
318          0.46090               0.11450          0.3135   
..               ...                   ...             ...   
371          0.13620               0.08178         

In [70]:
y_train

Result
176       1
260       0
206       1
319       1
519       1
..      ...
59        1
23        0
264       0
327       1
135       0

[369 rows x 1 columns]

In [71]:
y_test

Result
114       1
282       0
252       0
130       1
318       1
..      ...
371       1
47        0
489       0
220       1
423       1

[200 rows x 1 columns]

In [72]:
#Algorithm: LOgistic Regression

#parameters:
#1. penalty - to add L2 penalty or L1 penalty
#2. C - Inverse of regularisation strength.(1/lambda)
#3. class_weight - for not balanced dataset. Can apply weights to class.

#learn by 2 ways -probablistic and geometry.


In [74]:
params = [{'C': [1, 5, 10]}, {'max_iter':[100, 150]}]

#here, playing with two parameters(C, max_iter)
#max_iter - specifically for grid search CV


In [76]:
#Defining model:

model1=LogisticRegression(C=100,max_iter=100) # default values
#later on can apply gridsearch CV

model=GridSearchCV(model1, param_grid=params, scoring='f1', cv=5)
#since, it is classification problem. not pretty sure that whether 
#true+ve or true-ve is imporatnt

#Can use f1 scoring(It is basically parametric term-nothing but performance metrics)


In [77]:
#So, THis is the model with respect to GridSearchCV
model

GridSearchCV(cv=5, estimator=LogisticRegression(C=100),
             param_grid=[{'C': [1, 5, 10]}, {'max_iter': [100, 150]}],
             scoring='f1')

In [78]:
model.fit(X_train, y_train)

/opt/anaconda3/lib/python3.12/site-packages/sklearn/utils/validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/opt/anaconda3/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/utils/validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example

GridSearchCV(cv=5, estimator=LogisticRegression(C=100),
             param_grid=[{'C': [1, 5, 10]}, {'max_iter': [100, 150]}],
             scoring='f1')

In [79]:
model.best_params_

{'C': 10}

In [80]:
model.best_score_

0.9628721427611344

In [81]:
#96%

#still testing with test data

y_pred=model.predict(X_test)

In [82]:
y_pred

array([1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0,
       1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 1])

In [85]:
#Can apply confusion matrix after getting prediction values:

from sklearn.metrics import confusion_matrix,classification_report,accuracy_score


In [86]:
#confusion matrix with respect to predicted and actual values

confusion_matrix(y_test,y_pred)

array([[ 69,   2],
       [  8, 121]])

In [88]:
accuracy_score(y_test, y_pred)

#this gives accuracy for the model.

0.95

In [89]:
#for detailed precision, recall and all the scores:

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.90      0.97      0.93        71
           1       0.98      0.94      0.96       129

    accuracy                           0.95       200
   macro avg       0.94      0.95      0.95       200
weighted avg       0.95      0.95      0.95       200



In [ ]:
#Since, this is balanced dataset-the performance will be best.